### PR TITLE
Improve String's growth strategy

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Oof! Well there’s a classic space/time tradeoff if I ever saw one

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
That looks much better. I think we do a lot less element by element construction of Strings.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
The better benchmark results may just reflect benchmark coverage rather than true perf impact. I don't know if there's a strong argument that continuous appending (without pre-reservation) is more critical a use case than keeping memory use lower. We could have a 4x growth strategy and see benchmark wins too, so why should 2x be the true natural state?

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

File: stdlib/public/core/StringStorage.swift [hunk 1]
```diff
@@ -121,8 +121,13 @@ private func determineCodeUnitCapacity(_ desiredCapacity: Int) -> Int {
   let minCap = 1 + Swift.max(desiredCapacity, _SmallString.capacity)
   _internalInvariant(minCap < 0x1_0000_0000_0000, "max 48-bit length")
 
+#if _runtime(_ObjC)
+  let storageHeaderSize = 32
+  let capacity = malloc_good_size(minCap + storageHeaderSize)
```
Comment (milseman) [inline stdlib/public/core/StringStorage.swift]: We will then ask for another 8 bytes for the pointer, so wouldn't this under-utilize the bucket?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**